# D - Fiche 8 - Programmation fonctionnelle CORRECTION

# A. Effet de bord
# AFVM1 :  
Voici une fonction

    def doubler_tout(L):
        """procédure qui double chaque élément d'une liste de flottants"""
        for i in range(len(L)):
            L[i]=2*L[i]
        print(L)
        
1. Sans exécuter le script, deviner l'effet des deux appels `doubler_tout([1,4.5,8,7])` identiques successifs.<br>


In [1]:
def doubler_tout(L):
        """procédure qui double chaque élément d'une liste de flottants"""
        for i in range(len(L)):
            L[i]=2*L[i]
        print(L)
doubler_tout([1,4.5,8,7])
doubler_tout([1,4.5,8,7])

[2, 9.0, 16, 14]
[2, 9.0, 16, 14]


2. Une fonction est dite à effet de bord si elle modifie un état autre que sa valeur de retour. 

La fonction `doubler_tout` est-elle une fonction à effet de bord ?

**Oui, car la liste L est modifiée dans la fonction.**

3. Exécuter le code suivant. Que remarquez-vous ? Expliquez pourquoi.

        L = [1,2,3]
        doubler_tout(L)
        doubler_tout(L)
        print(L)

In [4]:
L = [1,2,3]
doubler_tout(L) # modification de L
doubler_tout(L) # Avec, ces 2 appels identiques, nous n'avons pas le même affichage 
print(L)

[2, 4, 6]
[4, 8, 12]
[4, 8, 12]


# B. Version itératives et récursives d'un même algorithme

# AFVM2 :  
1. On donne la fonction `fibo_rec` (suite de Fibonacci) suivante

    def fibo_rec(n):
        """Suite de Fibonacci, version récursive"""
        if n <= 1:
            return n
        else:
            return fibo_rec(n - 1) + fibo_rec(n - 2)
        
Calculer à la main `fibo_rec(4)`

`fibo_rec(4)` a pour valeur `fibo_rec(3) + fibo_rec(2)`
* avec `fibo_rec(3)` qui a pour valeur `fibo_rec(2) + fibo_rec(1)` soit `fibo_rec(2) + 1` soit `1 + 1 = 2` car 
* `fibo_rec(2)` qui a pour valeur `fibo_rec(1) + fibo_rec(0)` soit `1 + 0`

donc `fibo_rec(4)` a pour valeur `2 + 1 = 3`

2. Donner une version itérative de `fibo_rec`.

In [22]:
def fibo_iter(n):
    """TODO Suite de Fibonacci, version itérative"""
    f0 = 0
    f1 = 1
    f2 = 1
    if n <= 1:
            return n
    else:
        for i in range(n-2):
            f0, f1, f2 = f1, f2, f1+f2
    return f2
    
fibo_iter(5)       

5

# AFVM3 : 

On donne la fonction factorielle suivante :


    def fact_iter(n):
        """Factorielle, version itérative"""
        r = 1
        for i in range(1, n + 1):
            r = r*i
        return r

Donner une version recursive de `fact_iter`.

In [24]:
def fact_rec(n):
    """Factorielle, version recursive"""
    if n == 0:
        return 1
    else : 
        return n*fact_rec(n-1)
    
assert fact_rec(3)==3*2*1
assert fact_rec(4)==4*3*2*1

# C. Lambda fonctions

# AFVM4 : 
Qu'affiche le programme suivant ? Le faire de tête.

    print((lambda x : x//2 == 0)(5))
    print((lambda x,y:y**x)(2,3))
    print(list(map(lambda x : x+2, range(10))))

In [25]:
def f(x):
    return x//2 == 0
f(5)

False

In [26]:
def g(x,y):
    return y**x
g(2,3)

9

In [47]:
def h(x) : 
    return x+2
#la fonction map appelle la fonction h sur tous les élements de l'itérable range(10)
# attention ne renvoie pas une liste, on doit donc utiliser la fonction list() de conversion
print([h(elt) for elt in range(10)])

[2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


# AFVM5 :
1. De tête, qu'affiche le programme suivant ?

        def calcul(op,t): 
            v = t[0]
            for i in range(1,len(t)): 
                v = op(v,t[i])
            return v
    
        t = [1, 2, 3, 4, 5]
        print(calcul((lambda x, y: x+y),t))
        print(calcul((lambda x, y: x*y),t))
        print(calcul((lambda x, y: max(x,y)),t))


In [48]:
def calcul(op,t): 
    v = t[0]
    for i in range(1,len(t)): 
        v = op(v,t[i])
    return v

t = [1, 2, 3, 4, 5]
print(calcul((lambda x, y: x+y),t)) # somme des élements du tableau t
print(calcul((lambda x, y: x*y),t)) # produit des élements du tableau t
print(calcul((lambda x, y: max(x,y)),t)) # maximum des élements du tableau t

15
120
5


2. Écrire une fonction nommée `calc` qui calcule le résultat de $f(x_1) ⊕ f(x_2) ⊕ ... ⊕ f(x_n)$
sur les éléments $x_1$,...,$x_n$ d’un tableau, pour une fonction $f$ et une opération $⊕$ quelconques.

In [49]:
def calc(f,op,t): 
    v = f(t[0])
    for i in range(1,len(t)): 
        v = op(v,f(t[i]))
    return v

def calc2(f,op,t):
    return calcul(op, list(map(f,t)))

3. Appliquer cette nouvelle fonction `calc` pour obtenir la représentation du contenu d’un tableau contenant des nombres entiers sous la forme d’une chaîne de caractères de la forme "1, 2, 3, 4" c’est-à-dire où les éléments sont séparés par des virgules.

In [50]:
f = lambda x : str(x)
op = lambda x,y : x + ", "+y
t = [1, 2, 3, 4]
print(calc(f,op,t))
print(calc2(f,op,t))

1, 2, 3, 4
1, 2, 3, 4


# AFVM6 :
On souhaite écrire une fonction `double(f)` qui reçoit une fonction `f` en argument et renvoie une fonction qui applique deux fois de suite la fonction  `f` à son argument.
1. Que vaut `double(lambda x : x*x)(2)` ?
1. Que vaut `double(double(lambda x : x*x))(2)` ?
2. Ecrire la fonction `double(f)` et vérifier vos résultats  précédents.

In [37]:
f = lambda x : x*x
print(f(f(2))) # double(lambda x : x*x)(2)
print(f(f(f(f(2))))) # double(double(lambda x : x*x))(2)

16
65536


In [39]:
def double(f) : 
    """prend en argument une fonction et renvoie une fonction"""
    return lambda x : f(f(x))

print(double(lambda x : x*x)(2))
print(double(double(lambda x : x*x))(2))

16
65536


# AFVM7 :  Programmation fonctionnelle avec des variantes de `forall`

On va définir la fonction suivante `forall`, de plusieurs façons différentes

La fonction `forall` prend en paramètres
* une fonction `pred` (appelée prédicat)
* une liste `l`
`forall(l)` renvoie True si et seulement si tous les `x` dans `l` sont tels que `pred(x)==True`

1. Qu'affiche le code ci dessous ?

        ex1 = [4, 4, 2, 0, 1]
        ex2 = []
        ex3 = [4, 4, 2, 0, 8]


        def pair(x):
            """Un test de parité"""
            return x % 2 == 0
        
        print(forall(pair, ex1))
        print(forall(pair, ex2))
        print(forall(pair, ex3))


In [ ]:
False
True
True

2. Définir `forall` (version boucle `for`) : 

In [52]:
def forall_for(pred, l):
    """TODO : une version de forall avec une boucle for """
    resultat = True
    for x in l :
        if not pred(x):
            resultat = False
    return resultat

ex1 = [4, 4, 2, 0, 1]
ex2 = []
ex3 = [4, 4, 2, 0, 8]

def pair(x):
    """Un test de parité"""
    return x % 2 == 0
        
print(forall_for(pair, ex1))
print(forall_for(pair, ex2))
print(forall_for(pair, ex3))

False
True
True


3. Définir `forall` (version récursive avec les listes). 
Pour cela, on utilisera les fonctions suivantes : 

In [54]:
def empty(l):
    """Retourne True ssi la liste est vide"""
    return len(l) == 0


def head(l):
    """Retourne la tête de la liste, son 1er élément, et None si la liste est vide"""
    if empty(l):
        return None
    return l[0]


def tail(l):
    """Retourne la liste privée de son 1er élément, et None si la liste est vide"""
    if empty(l):
        return None
    return l[1:]


def forall_funct(pred, l):
    """TODO : une version récursive de forall, plus fonctionnelle, utilisant UNIQUEMENT
    empty, head et tail, sans boucle for ni affectation. La fonction sera récursive"""
    if empty(l):
        return True
    else : 
        return pred(head(l)) and forall_funct(pred, tail(l))

ex1 = [4, 4, 2, 0, 1]
ex2 = []
ex3 = [4, 4, 2, 0, 8]

def pair(x):
    """Un test de parité"""
    return x % 2 == 0
        
print(forall_funct(pair, ex1))
print(forall_funct(pair, ex2))
print(forall_funct(pair, ex3))

False
True
True


4. Définir `forall` (version avec `map`)

In [55]:
def forall_map_all(pred, l):
    """TODO : une version de forall utilisant les fonctions map et all de la bibliothèque standard
    voir https://docs.python.org/3/library/functions.html"""
    return all(map(pred, l))

ex1 = [4, 4, 2, 0, 1]
ex2 = []
ex3 = [4, 4, 2, 0, 8]

def pair(x):
    """Un test de parité"""
    return x % 2 == 0
        
print(forall_map_all(pair, ex1))
print(forall_map_all(pair, ex2))
print(forall_map_all(pair, ex3))

False
True
True


5. Définir `forall` (version avec `filter`)

In [57]:
list(filter(pair, ex1)) #equivalent to the generator expression (item for item in iterable if function(item)

[4, 4, 2, 0]

In [58]:
def forall_filter(pred, l):
    """TODO : une version de forall utilisant les fonctions standards filter et len. On pourra avoir
    besoin de transformer le résultat intermédiaire avec list()"""
    return len(list(filter(pred, l)))==len(l) 

ex1 = [4, 4, 2, 0, 1]
ex2 = []
ex3 = [4, 4, 2, 0, 8]

def pair(x):
    """Un test de parité"""
    return x % 2 == 0
        
print(forall_filter(pair, ex1))
print(forall_filter(pair, ex2))
print(forall_filter(pair, ex3))

False
True
True


6. Définir `forall` (version avec compréhension de listes)

In [59]:
def forall_comprehension(pred, l):
    """TODO : une version de forall utilisant la compréhension de liste au lieu de map ou filter"""
    return len([elt for elt in l if pred(elt)])==len(l) 

print(forall_comprehension(pair, ex1))
print(forall_comprehension(pair, ex2))
print(forall_comprehension(pair, ex3))

False
True
True


7. Définir `forall`  (version avec `reduce`)

La fonction standard `reduce` du module `functools` (appelée aussi `fold`) prend trois paramètres
* une fonction `f`  de type A * B -> A
* une liste `l` de type [A]
* une valeur initiale `x0` de type A

Pour `l = [x1, x2, ..., xn]`, `reduce(f, l, x0)` calcule`f(...(f(f(x0, x1), x2)...), xn)`

Par exemple, `reduce(lambda acc, x : acc * x, range(1,7), 1)` calcule `((((((1*1)*2)*3)*4)*5)*6) = 6! = 120`

https://docs.python.org/3/library/functools.html#functools.reduce

In [60]:
import functools

def forall_reduce(pred, l):
    """TODO : une version de forall utilisant UNIQUEMENT la fonction standard functools.reduce"""
    return functools.reduce(lambda x,y : x and pred(y),l,True)

print(forall_reduce(pair, ex1))
print(forall_reduce(pair, ex2))
print(forall_reduce(pair, ex3))

False
True
True
